src/Microsoft.OData.Core/UriParser/SemanticAst/SelectExpandClause.cs

```diff
@@ -49,7 +49,7 @@ public IEnumerable<SelectItem> SelectedItems
             get
             {
                 ////Debug.Assert(!this.usedInternalLegacyConstructor || this.selectedItems != null, "You cannot get the list of selected items until processing is complete.");
-                return this.selectedItems.AsEnumerable();
+                return new ReadOnlyCollection<SelectItem>(this.selectedItems).AsEnumerable();

```

Comment:
@ificator I would probably have the old code here as it will reduce the amout of allocations for cases where reading the selected items happens alot.

Reply:
I debated what we wanted here - I could just return `this.selectedItems` but that would be someone could theoretically cast it back to a `IList<SelectItem>` and modify our collection. Another option would be to have to allocate the `ReadOnlyCollection<SelectItem>` once - we should be able to modify the underlying collection independently of the read-only instance.

I ended up going with this approach because I rationalized the allocation of the `ReadOnlyCollection<SelectItem>` is pretty cheap (it's just a wrapper), but I'm willing to update to one of the other options (or something else?) if there's a preference.

Reply:
@ificator -- That seems reasonable to me.